### ISSUE TITLE
Sonoma: Uebersicht App does not not show widget list any more (although all widgets were launched properly)

### ISSUE THREAD

**Issue by maveice** [felixhageloh/uebersicht#516]
I am (was) enjoying your wonderful App "Übersicht" such a long time.

Some days ago, right after upgrading maOS to Sonoma, Uebersicht starts up as usual. However, I cannot see any "Uebersicht" widget-items any more (in dropdown from the MenuBar) which I can assign to my multiple screens or even edit ... (pls. see attachment)

Help of the Community is very much appreciated.
Many thanks thanks in advance!

![Widget list missing](https://github.com/felixhageloh/uebersicht/assets/19212442/3a3c82fd-36c5-48a3-a803-37d5bfd17735)

**Comment by robert914** [comment]
I can confirm that I'm seeing this same behavior on Sonoma.

**Comment by felixhageloh** [comment]
I have not been able to reproduce so far, but I have a theory what is going on. Could one of you try this release and see if the issue is fixed https://tracesof.net/uebersicht/releases/Uebersicht-1.6.72.app.zip?

**Comment by maveice** [comment]
Thank you for your swift action, Felix. 
I checked above version, however there is unfortunately no effect.

**Comment by robert914** [comment]
Ok, the *email* I got from GitHub said to download version *1.6.72* which is what I was using, and it had no change. However, when I came here to post a comment, I noticed you updated the link to *1.6.73*. So I re-downloaded that version, and now mine is working correctly on Sonoma! Thank you Felix!

**Comment by lucaorio** [comment]
I have the same problem and can no longer refresh the widgets programmatically!
`osascript -e "tell application id \"tracesOf.Uebersicht\" to refresh widget id \"widget\""`

**Comment by felixhageloh** [comment]
🤔 alright so the search continues. Could one of you record a log when Übersicht starts? Here is a guid on how to achieve that (check the 'Pinpoint and share' section): https://support.apple.com/guide/console/welcome/mac

**Comment by maveice** [comment]
Sure - here we go.
[debug.txt](https://github.com/felixhageloh/uebersicht/files/12801360/debug.txt)

**Comment by robert914** [comment]
Here's mine
[debug.txt](https://github.com/felixhageloh/uebersicht/files/12804685/debug.txt)

**Comment by spmfte** [comment]
## Temporary Solution

I've encountered the same problem after updating to macOS Sonoma, where widgets from Übersicht no longer appear in the menu bar dropdown. Here's a detailed walk-through of my diagnostic process and the observations:

### **Environment**:
- macOS Version: `14.0`
- Übersicht Version: `1.6`
- yabai Version: `yabai-v5.0.9`

### **Initial Solutions**:
1. I tried using the command:
   `yabai -m signal --add event=space_changed action="osascript -e 'tell application id \"tracesOf.Uebersicht\" to refresh'"`
   This did trigger a refresh for Übersicht when changing spaces using yabai and skhd. However, I observed a slight lag with this method, making it not ideal for my use.

2. Running the command `osascript -e 'tell application id \"tracesOf.Uebersicht\" to refresh'` manually did refresh Übersicht. But, for reasons yet to be determined, this refresh wasn't automatic when switching spaces using yabai and skhd, which was the desired behavior.

### **Further Diagnostic Steps**:

1. Checked logs for errors relating to Übersicht using:
   `log show --predicate 'process == "tracesOf.Uebersicht"' --info --last 10m`
   No significant errors were found.

2. Ensured that Übersicht, yabai, and skhd were running without issues using:
   `pgrep -fl tracesOf.Uebersicht`

3. Checked for sandboxing issues related to Übersicht:
   `log show --predicate 'subsystem == "com.apple.sandbox.reporting" AND process == "tracesOf.Uebersicht"' --info --last 1d`
   No issues found.

4. Checked the DiagnosticReports directory for crash reports pertaining to Übersicht.

### **Temporary Solution**:
Despite these diagnostics, I found that manually restarting Übersicht offers a temporary reprieve from the issue. This manual refresh can serve as a workaround until we identify a permanent solution.

**Comment by maveice** [comment]
On my system even a manual restart of the App doesn't work.

**Comment by felixhageloh** [comment]
Thanks @spmfte for the detailed report. I had some time this weekend to look through the other logs you all shared here and I could not find any obvious errors either. 

I still think it might be a race condition - although with version 1.6.73 I would have expected that to be fixed - mostly because there are no errors and widgets are actually showing on the desktop for all of you. 

I should have some time this week to look into this some more!

**Comment by felixhageloh** [comment]
One thing I was wondering: if you add a new widget or move an existing widget out of the widget folder and back again, does it show up in the menu?

**Comment by lucaorio** [comment]
Thanks, @felixhageloh, for taking a look at this! An interesting situation here:

This prints nothing while Übersicht is running:
```shell
osascript -e "tell application id \"tracesOf.Uebersicht\" to every widget"
# nothing printed
```

If I duplicate an existing widget, all of them appear!
```shell
cp -r ~/Library/Application\ Support/Übersicht/widgets/nero ~/Library/Application\ Support/Übersicht/widgets/nero-copy
osascript -e "tell application id \"tracesOf.Uebersicht\" to every widget"
«class » id nero-yabai-jsx, «class » id nero_copy-yabai-jsx 
```

But if I restart Übersicht, the widgets are not recognized again:
```shell
killall Übersicht
open /Applications/Übersicht.app
osascript -e "tell application id \"tracesOf.Uebersicht\" to every widget"
# nothing printed
```

**Comment by marconetto** [comment]
Thanks @felixhageloh and @lucaorio. Indeed, copying and existing widget, and removing it, makes things work normally.

**Comment by maveice** [comment]
confirmed, too.

**Comment by felixhageloh** [comment]
Alright thanks for the info! I think I am beginning to understand the issue. I have a new version here:
https://tracesof.net/uebersicht/releases/Uebersicht-1.6.74.app.zip

can you try it again and see if it fixes the problem?

**Comment by maveice** [comment]
Hi Felix, thank you - however, it does not solve the problem.
If the App is started and you then move and re-move some parts (files or dirs) from/to the widgets folder(s), everything works as expected. If you just (re)launch the App, we unfortunately still have the known situation.

**Comment by felixhageloh** [comment]
Does ‘refresh all widgets’ fix the issue now? I also changed how it works

**Comment by maveice** [comment]
no

**Comment by felixhageloh** [comment]
🤔that’s a pity. Can you share the console log with this new version as well? I added some extra logging that might shed more light on the issue

**Comment by maveice** [comment]
Here we go!

[console.txt](https://github.com/felixhageloh/uebersicht/files/12855944/console.txt)

**Comment by felixhageloh** [comment]
thanks @maveice! So it looks like it finds all your widgets correctly before the app fetches them. So it is not a race condition as I first thought - or at least not the race condition I had in mind. 
 
I am beginning to wonder whether it is a specific widget that is breaking. When you mentioned that things start working when you remove and add a widget again, did you mean that all widgets appear again or just the one that you just added?

**Comment by maveice** [comment]
It doesn't matter whih one you take out and then put back. The effect is same: For the moment it is working (all widgets there again) and after quitting the app and relaunching nothing is there anymore.

**Comment by maveice** [comment]
Regarding your particular question above: Yes, ALL wigets are back at this moment.

**Comment by robert914** [comment]
@felixhageloh I see the same behavior as @maveice . Duplicating and removing an existing widget makes all re-appear in the menu.

**Comment by corporate-gadfly** [comment]
> thanks @maveice! It looks like it finds all your widgets correctly before the app fetches them. So it is not a race condition as I first thought - or at least not the race condition I had in mind.
> 
> I am beginning to wonder whether it is a specific widget that is breaking. When you mentioned that things start working when you remove and add a widget again, did you mean that all widgets appear again or just the one that you just added?

I don't think it is widget-dependent at all, as some of the other commenters have pointed out. In my case, I have a mix of coffee and js widgets:
![image](https://github.com/felixhageloh/uebersicht/assets/1200537/fc8a4e38-9d9a-4e28-83ae-9743b732c550)

**Comment by robert914** [comment]
Ok, I installed this latest version 1.6.75 and all my widgets appear in the menu now even after quitting and restarting multiple times. So you may have found it! I'll report back if I see anything different. Thanks!

**Comment by maveice** [comment]
Felix, as I can only access my Mac by the end of next week I will give my feedback by then. FYI

**Comment by 2oh1** [comment]
Here's my workaround for this issue.

As mentioned above, moving a widget fixes this issue until the next time you restart your Mac.  So, I wrote the following (admittedly crappy) Applescript to move a widget out of the widgets folder, and then move it back in:

`set OriginalSource to POSIX file "/Users/Me/Library/Application Support/Übersicht/widgets/Word-Clock.coffee"
set OriginalLocation to POSIX file "/Users/Me/Library/Application Support/Übersicht/widgets/"
set TempSource to POSIX file "/Users/Me/Library/Application Support/Übersicht/Word-Clock.coffee"
set TempLocation to POSIX file "/Users/Me/Library/Application Support/Übersicht/"



tell application "Finder"
	move OriginalSource to folder TempLocation with replacing
end tell

delay 1

tell application "Finder"
	move TempSource to folder OriginalLocation with replacing
end tell`

I set that Applescript to run on startup.  So, whenever I restart my Mac, Ubersicht fixes itself.

**Comment by felixhageloh** [comment]
@2oh1 does  the new version posted here https://github.com/felixhageloh/uebersicht/issues/516#issuecomment-1774632662 fix the issue for you?

**Comment by 2oh1** [comment]
Nope.  That didn't solve it for me.  The widgets didn't show up in the menu until I moved one out and then back in, even after downloading [Uebersicht-1.6.75.app.zip](http://tracesof.net/uebersicht/releases/Uebersicht-1.6.75.app.zip).

**Comment by robert914** [comment]
@felixhageloh I have to post a correction to my above post. I got so excited to test the new version that I mistakenly did the test on my Ventura-based Mac which of course works fine. So I went back and tried the new version on my Sonoma-based Mac and the issue *is* still there, sadly. Sorry for any confusion this may have caused. I can confirm that I'm still not getting widgets appearing on my Sonoma-based Mac like @2oh1 is also reporting. Sorry.

**Comment by felixhageloh** [comment]
no worries @robert914 :). In that case can you share the output from the Console app (details shared further up in this thread)? This version also includes some extra debug logging to understand the issue better. 

Same request for you @2oh1

**Comment by robert914** [comment]
@felixhageloh Ok, here's my console output for any:Übersicht:

[console.txt](https://github.com/felixhageloh/uebersicht/files/13189621/console.txt)

**Comment by felixhageloh** [comment]
ok I have another version - also more logging again, so if it doesn't work I'm have to ask for the console log again 🙏  http://tracesof.net/uebersicht/releases/Uebersicht-1.6.76.app.zip

**Comment by robert914** [comment]
@felixhageloh I'm away from my Sonoma-based Mac, so I won't be able to try this new version until sometime tomorrow. But I'll let you know how it goes and send an updated console log (if needed).

**Comment by robert914** [comment]
@felixhageloh Ok, I was able to install and test 1.6.76 on my Sonoma-based Mac, and it seems to be working now. The initial start showed all the widgets, and I even tried quitting and starting it again, and they are still showing correctly! I also installed it on my Ventura-based Mac, and it's still working there as well. So I think (from my testing) it appears you've fixed it. 👍

**Comment by felixhageloh** [comment]
Nice, finally 😄! I'll wait for one more confirmation tho before I celebrate 🎉

**Comment by lucaorio** [comment]
I confirm [Uebersicht-1.6.76.app.zip](http://tracesof.net/uebersicht/releases/Uebersicht-1.6.76.app.zip) addresses the issue for me too! Thank you, @felixhageloh! 🙏

**Comment by 2oh1** [comment]
Here's another confirmation.  [Uebersicht-1.6.76.app.zip](http://tracesof.net/uebersicht/releases/Uebersicht-1.6.76.app.zip) solved it for me too.  I'm running Mac OS Sonoma on an M1 Mac Studio.  Problem solved.

**Comment by maveice** [comment]
Next confirmation (MacBook Air M1 + Mac Studio M1, both on Sonoma) - I think, Felix, you got it.
Thank you very much!

**Comment by felixhageloh** [comment]
Thank you everyone for your patience and helping out with debugging! I released the fix as version 1.6.77, which is the same as 1.6.76 minus some debug logging.

**Comment by robert914** [comment]
@felixhageloh Thank you for your dedication to fixing the problem! We appreciate it!
